providers/harness/pkg/provider.go

```diff
@@ -0,0 +1,310 @@
+package harness
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/harness/ff-golang-server-sdk/client"
+	harness "github.com/harness/ff-golang-server-sdk/client"
+	"github.com/harness/ff-golang-server-sdk/evaluation"
+	"github.com/harness/ff-golang-server-sdk/types"
+	of "github.com/open-feature/go-sdk/pkg/openfeature"
+)
+
+const providerNotReady = "Provider not ready"
+const generalError = "general error"
+
+type Provider struct {
+	providerConfig ProviderConfig
+	harnessClient  *client.CfClient
+	status         of.State
+}
+
+func NewProvider(providerConfig ProviderConfig) (*Provider, error) {
+	provider := &Provider{
+		providerConfig: providerConfig,
+		status:         of.NotReadyState,
+	}
+	return provider, nil
+}
+
+func (p *Provider) Init(evaluationContext of.EvaluationContext) error {
+	harnessClient, err := harness.NewCfClient(p.providerConfig.SdkKey, p.providerConfig.Options...)
+	if err != nil {
+		p.status = of.ErrorState
+	} else {
+		p.status = of.ReadyState
+		p.harnessClient = harnessClient
+	}
+	return err
+}
+
+func (p *Provider) Status() of.State {
+	return p.status
+}
+
+func (p *Provider) Shutdown() {
+	p.harnessClient.Close()
+	p.status = of.NotReadyState
+}
+
+// provider does not have any hooks, returns empty slice
+func (p *Provider) Hooks() []of.Hook {
+	return []of.Hook{}
+}
+
+// Metadata returns value of Metadata (name of current service, exposed to openfeature sdk)
+func (p *Provider) Metadata() of.Metadata {
+	return of.Metadata{
+		Name: "harness",
+	}
+}
+
+func (p *Provider) BooleanEvaluation(ctx context.Context, flag string, defaultValue bool, evalCtx of.FlattenedContext) of.BoolResolutionDetail {
+	if p.status != of.ReadyState {
+		if p.status == of.NotReadyState {
+			return of.BoolResolutionDetail{
+				Value: defaultValue,
+				ProviderResolutionDetail: of.ProviderResolutionDetail{
+					ResolutionError: of.NewProviderNotReadyResolutionError(providerNotReady),
+					Reason:          of.ErrorReason,
+				},
+			}
+		} else {
+			return of.BoolResolutionDetail{
+				Value: defaultValue,
+				ProviderResolutionDetail: of.ProviderResolutionDetail{
+					ResolutionError: of.NewGeneralResolutionError(generalError),
+					Reason:          of.ErrorReason,
+				},
+			}
+		}
+	}
+
+	harnessTarget, err := toHarnessTarget(evalCtx)
+	if err != nil {
+		return of.BoolResolutionDetail{
+			Value: defaultValue,
+			ProviderResolutionDetail: of.ProviderResolutionDetail{
+				ResolutionError: of.NewInvalidContextResolutionError(err.Error()),
+				Reason:          of.ErrorReason,
+			},
+		}
+	}
+
+	res, _ := p.harnessClient.BoolVariation(flag, harnessTarget, defaultValue)
+	return of.BoolResolutionDetail{
+		Value:                    res,
+		ProviderResolutionDetail: of.ProviderResolutionDetail{},
+	}
+}
```

Comment:
You might be able to use generics to reduce some duplication in these functions - but I'm not very experienced with Go generics, so I could be wrong.

Reply:
I re-arranged it a bit, such that each evaluation is now more minimal.